numpy.random.generator should be preferred to numpy.random.random state

Wuodan opened the pull request:
I'm getting these warnings:
_numpy.random.Generator should be preferred to numpy.random.RandomState_
from SonarLint wherever numpy.random.RandomState is used.

This is the Sonar rule: https://rules.sonarsource.com/python/RSPEC-6711/

Fixes #40

pb4git commented on trackmania_rl/buffer_utilities.py:
```diff
@@ -77,9 +77,10 @@ def buffer_collate_function(batch):
     temporal_mini_race_current_time_actions = (
         np.abs(
-            np.random.randint(
+            rng.integers(
```
Given that rng is initialized just above, I believe we will always receive the same random numbers.


Obligatory xkcd: https://xkcd.com/221/

I'd prefer to have a single random number generator, initialized once at the beginning.

Wuodan replied:
I changed the code in the PR to create 1 RNG and then spawn independent child RNG from it as described in
https://blog.scientific-python.org/numpy/numpy-rng/
at the bottom.

Had to refactor buffer_utilities a bit to avoid bouncing adding an additional parameter everywhere. Added a class BufferUtil there to hold the RNG instead.

pb4git reviewed (changes requested):
I am unfamiliar with github's review process. I have commented the modification on line 83 of iqn.py, can you see this comment?

Also, just to check: do you run the code prior to submitting pull requests?

Wuodan committed:
Use a "numpy.random.Generator" here instead of this legacy function.

Init one RNG and spawn independent RNG from it.
Create class BufferUtil to preserve RNG

Fixes #40

Wuodan committed:
ruff formatting

with:
```
for file in $(git diff --name-only HEAD~1 HEAD); do
  echo "$file"
  ruff check --select I --fix "$file"
  ruff format "$file"
done
```

Wuodan commented:
> I am unfamiliar with github's review process. I have commented the modification on line 83 of iqn.py, can you see this comment?

Yes, I can.

> Also, just to check: do you run the code prior to submitting pull requests?

I went over the entire code and fixed many small things in about 20-30 commits in one branch.
The result of it was tested thoroughly.
But I'm not testing every single PR with just 1 commit and a tiny change - which was tested together with other changes -  for hours :-)

This one here was tested, not for hours but until something was written to disk plus some debugging in the running app.

Once I'm done with the easy PRs (a branch each, ruff format, some text in an issue) I want to merge all that locally again and will run more extensive test.

pb4git commented:
Not pulled yet, need to understand the wrapper around buffer utilities.

Wuodan commented:
The reason for the new class `BufferUtil` is that the random generator `rng` is used in many places in buffer_utilities.py and that file has quite a few functions which call each other and `rng` is only used by a few way down the chain.

So instead of adding `rng` to a bunch of functions, where it mostly would not be used but just be passed on, `BufferUtil` holds the `rng` and the affected functions were moved into `BufferUtil` as visible here (before and after structure):
<img width="658" alt="image" src="https://github.com/user-attachments/assets/4e30d80a-628a-49f7-a88d-306f3215625f">

`rng` is an instance variable of `BufferUtil` and does not have to be passed around.

`BufferUtil` itself is instantiated only once in learner_process.py:
`buffer_util = BufferUtil(rng)`
so there is only one instance of it and the overhead is very small.
